torch/nn/init.py

```diff
@@ -1,239 +1,394 @@
 import math
+import types
 import random
+import warnings
 
 import torch
 
+##########################
+# base initializer
+##########################
 
-def calculate_gain(nonlinearity, param=None):
-    """Return the recommended gain value for the given nonlinearity function.
-    The values are as follows:
 
-    ============ ==========================================
-    nonlinearity gain
-    ============ ==========================================
-    linear       :math:`1`
-    conv{1,2,3}d :math:`1`
-    sigmoid      :math:`1`
-    tanh         :math:`5 / 3`
-    relu         :math:`\sqrt{2}`
-    leaky_relu   :math:`\sqrt{2 / (1 + negative\_slope^2)}`
-    ============ ==========================================
+class Initializer(object):
+    """
+    Base class for all initializations.
+    """
+    def __new__(cls, *non_tensor_args, **non_tensor_kwargs):
+        first = None
+        if len(non_tensor_args) > 0:
+            first = non_tensor_args[0]
+        if first is None:
+            return object.__new__(cls)
+        if isinstance(first, (torch.Tensor, torch.autograd.Variable)):
+            return cls(*non_tensor_args[1:], **non_tensor_kwargs)(first)
+        else:
+            return object.__new__(cls)
+
+##########################
+# initializers
+##########################
+
+
+class Ones(Initializer):
+    r"""Fills the input Tensor with ones.
 
     Args:
-        nonlinearity: the nonlinear function (`nn.functional` name)
-        param: optional parameter for the nonlinear function
+        tensor: an n-dimensional `torch.Tensor`
 
     Examples:
-        >>> gain = nn.init.calculate_gain('leaky_relu')
+        >>> w = torch.Tensor(3, 5)
+        >>> nn.init.ones_(w)
+        >>> ones_init = nn.init.ones_()
+        >>> ones_init(w)
     """
-    linear_fns = ['linear', 'conv1d', 'conv2d', 'conv3d', 'conv_transpose1d', 'conv_transpose2d', 'conv_transpose3d']
-    if nonlinearity in linear_fns or nonlinearity == 'sigmoid':
-        return 1
-    elif nonlinearity == 'tanh':
-        return 5.0 / 3
-    elif nonlinearity == 'relu':
-        return math.sqrt(2.0)
-    elif nonlinearity == 'leaky_relu':
-        if param is None:
-            negative_slope = 0.01
-        elif not isinstance(param, bool) and isinstance(param, int) or isinstance(param, float):
-            # True/False are instances of int, hence check above
-            negative_slope = param
-        else:
-            raise ValueError("negative_slope {} not a valid number".format(param))
-        return math.sqrt(2.0 / (1 + negative_slope ** 2))
-    else:
-        raise ValueError("Unsupported nonlinearity {}".format(nonlinearity))
+    def __call__(self, tensor):
+        with torch.no_grad():
+            return tensor.fill_(1)
 
 
-def uniform(tensor, a=0, b=1):
-    """Fills the input Tensor or Variable with values drawn from the uniform
-    distribution :math:`U(a, b)`.
+class Zeros(Initializer):
```

Comment:
Can you please reorder those in the same way as they appeared in the original file? Right now it's hard to review them, because the diffs are really messed up